### DISCUSSION TITLE
K3s Cluster with MetalLB not routing external traffic to K3s (via L2)

### DISCUSSION THREAD

**Discussion by steddyman** [k3s-io/k3s#8202] [Q&A]
Hi
I have been trying for one week solid to get MetalLB working with K3s on Hetzner bare metal servers (Robot)
I believe I have everything setup correctly.  I have disabled serviceLB and even network policies and UFW for troubleshooting.
I am using the Hetzner vSwitch facility with both a private VLAN for the servers to talk to each other on 192.168.100.x/24 and a public /29 IP range for the LB on a 159.x.x.x range.
Both my ingress and services marked as LoadBalancer are receiving external IP addressed from my defined range via MetalLB
The IP's are accessible from any of the Hetzner nodes using ARPING to confirm a MAC is assigned by MetalLB.
The pods are also accessible from inside the cluster private IP range via CURL from another pod.
I have setup a LoadBalancer 'Service' (rather than Ingress) to rule out issues with Traefik Ingress controller which shows the exact same issue.
Using ARPING I can see the host MetalLB has assigned the host to handle the address correctly and it gets a MAC on the correct interface.
When I perform a CURL to the assigned external IP of the service remotely from my laptop, i just get connection timed out.
That same CURL to the Service external URL (via LB IP) works from any of the nodes.
I can run TCPDUMP -I interface_name.vlan external.ip.address and I can see the SYN packets being received from the remote laptop device, but no ACK packets sent and the traffic does not reach the PODS.
Oddly, if I run MTR --tcp -P port 159.x.x.x to test port access, I can see it reaches the network interface of my host, only if it is NOT LISTENING on that port.  If the port I probe is the one with a service attached, the last hop showing the host name is missing from the output.
The pod logs show it does not receive the request externally across the Internet, but does locally from another Node or POD in the cluster.
There are no errors in the Speaker logs for the service.
Again, UFW and Network policies are currently disabled to troubleshoot.
I am up against the limits of my knowledge now.  I've spoken to the MetalLB group and they tell me MetalLB has done its job and the issue is now with the CNI in K3s.  I am using the default flannel VXLAN.
Can anyone please provide guidance as to what could be happening here please?

I have also tried manually adding one of the LB IP's to a host and when I do that, I can ping and reach it from my remote laptop just fine, so the route has been properly setup and advertised on the Hetzner end.

Thanks

**Comment by withinboredom** [comment]
Did you ever manage to solve this? I'm running into exactly this issue. All I know for sure is that flannel is dropping the packets.

For now, I'll just record my observations while I attempt to diagnose the issue.

1. I have three ifaces (1 physical, two vlans): an "external" vlan that has external ips, and an "internal" vlan for communicating between nodes (using host-gw mode of flannel).
2. Warning: having an internal IP on the "external" vlan means traffic can't be routed correctly -- I opened an issue with Hetzner support as traffic from an internal IP to the external IP in the same vlan should be routable.
3. I can see the SYN packets arriving on the "external" iface, but not going anywhere. I do see some routing to cni0 with ipv6 ips, but responses aren't going anywhere. Likely need to enable ipv6 masquerading?
4. Let's get ipv4 working first, before dealing with ipv6.
5. It's almost like flannel isn't seeing the packets. I'm not sure which iface flannel is bound to, but let's force it to be bound to the physical interface. That _should_, hopefully let it see all traffic. The physical interface should see the vlan routes and send the packets through the right interface.
6. It looks like there is a slightly undocumented argument: `flannel-iface` that can do this. 
7. This appears to have no effect.
8. I note that ipv6 connectivity to the internet is broken. Maybe I should go ahead and enable ipv6-masq
9. now we can connect to the internet from the pods via ipv6 (but still not the other way around)
10. Just revisiting the ipv6 side now, since I am at a dead-end for the ipv4 side. In this case, packets are routed all the way through, from internet to pod, and back again. However, they never return to me. I suspect I'll need to tcpdump to a file and analyze in wireshark to look at the layer2 headers to know what is going on for sure. Here we go...
11. From looking at this, it looks like I should be receiving ipv6 packets.
12. I am not. Even after directly connecting to the internet to ensure I didn't have some firewall on my own router blocking things.
13. ah, the return packets are only being emitted in the cni0 interface, they never reach the physical interface.
14. This is starting to look like some flannel shenanigans, indeed. Flannel doesn't know how to handle the external ips... from the looks of things. 
15. Oh... well, this will never work with Flannel: https://github.com/flannel-io/flannel/issues/392

Off to find a CNI that can support this use-case... probably cillium.

**Reply by withinboredom** [reply]
Cillium works almost out of the box with this configuration.

Updated cluster with the following config:

```yaml
write-kubeconfig-mode: "0644"
disable:
  - local-storage
  - servicelb
  - traefik
  - kube-proxy

# networking options
flannel-backend: none
disable-network-policy: true
cluster-cidr: 10.42.0.0/16,fd28:cae1:9e6c::/56
service-cidr: 10.43.0.0/16,fd28:cae1:9e6c::/112

advertise-address: 172.17.100.3
node-ip: 172.17.100.3,fd28:cae1:9e6c::3
```

And here is the cilium install command I gave:

```
cilium install \
--namespace kube-system \
--set tunnel=geneve \
--set kubeProxyReplacement=true \
--set loadBalancer.mode=dsr \
--set loadBalancer.dsrDispatch=geneve \
--set k8sServiceHost=172.17.100.3 \ # todo: replace with DNS name when more servers join
--set k8sServicePort=6443 \
--set ipv6.enabled=true \
--set ipam.operator.clusterPoolIPv6PodCIDRList=fd28:cae1:9e6c::/112
```

After a reboot (to flush out all the left over iptables and networking shenanigans), everything works again.
